c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1638,189 +1648,148 @@ class DeviceCachingAllocator {
     return block;
   }
 
-  // Insert "free marker" (empty nodes) into the CUDA graph for all streams that
-  // have used the block, including the allocation stream. These nodes mark the
-  // last use of the block in the capture graph. Returns a vector of the
-  // inserted nodes, or an empty vector if any stream is not capturing.
-  std::vector<cudaGraphNode_t> insert_free_marker(Block* block) {
-    std::vector<cudaGraphNode_t> empty_nodes;
-
-    auto try_add_empty_node = [&](cudaStream_t stream) -> bool {
-      cudaStreamCaptureStatus status{};
-      cudaGraph_t graph{};
-      const cudaGraphNode_t* deps = nullptr;
-      size_t num_deps = 0;
+  struct CaptureInfo {
+    cudaGraph_t graph{};
+    const cudaGraphNode_t* terminals{nullptr};
+    size_t num_terminals{0};
+    cudaStreamCaptureStatus status{cudaStreamCaptureStatusNone};
+  };
+
+  inline CaptureInfo stream_get_capture_info(cudaStream_t stream) {
+    CaptureInfo info{};
 #if (defined(CUDA_VERSION) && CUDA_VERSION >= 13000)
-      C10_CUDA_CHECK(cudaStreamGetCaptureInfo(
-          stream, &status, nullptr, &graph, &deps, nullptr, &num_deps));
+    C10_CUDA_CHECK(cudaStreamGetCaptureInfo(
+        stream,
+        &info.status,
+        nullptr,
+        &info.graph,
```

Comment:
Actually, I think you should switch from using `graph` as a map key to using unique id, `graph` is just a pointer, after hashing its not guaranteed to be unique, and what you want is a unique identifier for a graph, see e.g. how it's currently used in `capture_begin` https://github.com/pytorch/pytorch/blob/cf94cadbeee31a4d1d46a57f11bce7c9fd1cebc0/aten/src/ATen/cuda/CUDAGraph.cpp#L103